Security privileges

SaschaPeukert opened the pull request.

SaschaPeukert committed:
More work on privileges

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/grant-deny-syntax.asciidoc:
```diff
@@ -52,7 +40,8 @@ allows the specified entities to be found
 allows the specified properties to be read on the found entities.
-Note that if the way the entity is found requires reading the property, it will not be found.
+Note that granting property READ access does not imply that the entities with that property can be found.
+For example if there is also a DENY on it, the entity will not be found.
```
Or missing TRAVERSE?

SaschaPeukert replied:
Thats why i was saying "for example". I was not meaning to give a complete list, just an example

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityPrivilegesTest.scala:
```diff
@@ -34,9 +36,21 @@ class SecurityPrivilegesTest extends DocumentingTest with QueryStatisticsTestSup
+        resultTable()
+      }
+
+      p("Available privileges for a particular user can be seen using `SHOW USER $role PRIVILEGES`.")
```
'SHOW USER **$role** PRIVILEGES' Should probably be user

Hunterness replied:
shoudl there even be a $ here (and for roles)? It is not a parameter that should be supplied...

Hunterness replied:
I know that the example queries for this in the role and user management sections use `SHOW USER/ROLE name PRIVILEGES`

SaschaPeukert committed:
Covering all of the other privileges

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityPrivilegesTest.scala:
```diff
@@ -16,16 +16,18 @@ class SecurityPrivilegesTest extends DocumentingTest with QueryStatisticsTestSup
         |The user will be able to access the resource if they have a grant (whitelist) and do not have a deny (blacklist) relevant to that resource.
-        |If there are no read-privileges provided at all, then the user will be denied access to the entire graph, and this will generate an error.
+        |If a user was not provided with the access privilege then access to the entire graph will be denied.
         |All other combinations of GRANT and DENY will result in the matching subgraph being visible.
```
Sounds slightly weird to say `All other combinations of GRANT and DENY` when neither was mentioned just before.

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityPrivilegesTest.scala:
```diff
@@ -34,12 +36,30 @@ class SecurityPrivilegesTest extends DocumentingTest with QueryStatisticsTestSup
     }
     section("Listing privileges", "administration-security-subgraph-show") {
-      p("Available privileges for a particular role can be seen using `SHOW  PRIVILEGES`.")
+      p("Available privileges for all roles can be seen using `SHOW  PRIVILEGES`.")
```
Two spaces in command

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityPrivilegesTest.scala:
```diff
@@ -50,19 +70,60 @@ class SecurityPrivilegesTest extends DocumentingTest with QueryStatisticsTestSup
+      p("Privileges that were granted or denied earlier can be revoked using the `REVOKE` command. ")
+      p("include::grant-match-syntax.asciidoc[]")
+
+      p("Please note that `REVOKE MATCH` is not allowed.")
```
could be added as a note instead of a paragraph

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/grant-deny-syntax.asciidoc:
```diff
@@ -56,9 +56,6 @@ this combines both TRAVERSE and READ allowing an entity to be found and its prop
-//TODO:
-** `ALL OTHER PRIVILEGES +{*}+`
-+
 TODO: Add all the other privileges here, there are quite a few!
```
This needs to be added

Hunterness replied:
Are there more graph privileges left? Or are all others part of the admin section?

craigtaverner replied:
I think they are all part of the admin section. Also the `ACCESS` privilege above should be moved to admin.

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityPrivilegesTest.scala:
```diff
@@ -42,22 +42,42 @@ class SecurityPrivilegesTest extends DocumentingTest with QueryStatisticsTestSup
+      p("include::deny-access-syntax.asciidoc[]")
 
+      p("For example, denying the ability to access to the database `neo4j` to the role `regularUser` is done like the following query.")
+      query("DENY WRITE { * } ON GRAPH neo4j TO regularUser", ResultAssertions((r) => {
```
This says `deny write`

Hunterness replied:
Also the ACCESS syntax was wrong both here and in the syntax file

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityPrivilegesTest.scala:
```diff
@@ -70,10 +90,6 @@ class SecurityPrivilegesTest extends DocumentingTest with QueryStatisticsTestSup
         p("Nothing is returned from this query, except the count of system database changes made.")
         resultTable()
       }
-      p("The privileges granted to the `regularUser` role will appear on the list provided by `SHOW ROLE $role PRIVILEGES`.")
```
Why is these removed? (also line 87-90)

Hunterness replied:
I think these examples are nice to have (also for the read and match versions)

Hunterness replied:
From Sascha: Since the list will have to be everywhere and will contain all past privileges it might be weird if you only look at one section

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityPrivilegesTest.scala:
```diff
@@ -84,33 +100,109 @@ class SecurityPrivilegesTest extends DocumentingTest with QueryStatisticsTestSup
     }
     section("The MATCH privilege", "administration-security-subgraph-match") {
       p("As a shorthand for `TRAVERSE` and `READ`, users can be granted the right to find and do property reads on nodes and relationships using the `GRANT MATCH` privilege. ")
+      p("Please note that `REVOKE MATCH` is not allowed.")
```
Would like to have this as a note instead of a paragraph (and maybe last in the section?)

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityPrivilegesTest.scala:
```diff
@@ -84,33 +100,109 @@ class SecurityPrivilegesTest extends DocumentingTest with QueryStatisticsTestSup
+
+      p("""Please note that the effect of denying a `MATCH` privilege depends on whether concrete property keys are specified or a `*`.
+          |If you specify concrete property keys then `DENY MATCH` will only deny reading those properties. Finding the elements to traverse would still be allowed.
+          |If you specify `*`instead then both traversal of the element and all property reads will be disallowed.
```
missing space

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityPrivilegesTest.scala:
```diff
@@ -50,19 +90,132 @@ class SecurityPrivilegesTest extends DocumentingTest with QueryStatisticsTestSup
+      p("An example usage of the `REVOKE` command is given here:")
+      query("REVOKE GRANT TRAVERSE ON GRAPH neo4j NODES Post TO regularUser", ResultAssertions((r) => {
+        assertStats(r, systemUpdates = 1)
+      })){}
```
The empty body here makes sure we don't even check the assertion on statistics

Hunterness replied:
Use the new statsOnlyResultTable

Hunterness committed:
Fixed some of the review comments

Hunterness commented:
The ACCESS privilege should move, @craigtaverner will write the part it should move to so he took that on.

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ConstraintsTest.scala:
```diff
@@ -21,7 +21,7 @@ package org.neo4j.cypher.docgen
 
 import java.io.File
 
-import com.neo4j.dbms.api.EnterpriseDatabaseManagementServiceBuilder
```
Apparently not enough fixed, will try again

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityPrivilegesTest.scala:
```diff
@@ -16,30 +16,75 @@ class SecurityPrivilegesTest extends DocumentingTest with QueryStatisticsTestSup
         |The user will be able to access the resource if they have a grant (whitelist) and do not have a deny (blacklist) relevant to that resource.
-        |If there are no read-privileges provided at all, then the user will be denied access to the entire graph, and this will generate an error.
-        |All other combinations of GRANT and DENY will result in the matching subgraph being visible.
+        |All combinations of GRANT and DENY will result in the matching subgraph being visible.
```
This line is completely wrong. It needs the words `other` and `not`.

Hunterness replied:
Other then what, there is no talk about anything with grant and deny (just access)?

craigtaverner commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/grant-deny-syntax.asciidoc:
```diff
@@ -32,27 +32,20 @@ REVOKE graph-privilege ON GRAPH dbname entity FROM role
-*** `ELEMENTS` label (both nodes and relationships).
-** The label or type can be `+*+` which means all labels or types.
 * _graph-privilege_
+** `ACCESS`
```
This is _NOT_ a graph-privilege it is a database-privilege and should be moved to the admin section.

craigtaverner committed:
Resolved some PR review comments and made a start on admin privileges

In addition there is initial support for simplified results for stats-only queries (updating queries).
This must be matched by some compatible changes in the neo4j-manual-modeling, so be sure to get the latest version of that before building the entire manual.

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/admin-role-introduction.asciidoc:
```diff
@@ -0,0 +1,116 @@
++
+* _dbname_
+** The database to associate the privilege with.
+   Note that if you delete a database and create a new one with the same name, the new one will NOT have any of the privileges specifically assigned to the deleted graph.
```
`privileges specifically assigned to the deleted graph.` should be database not graph?

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/tooling/DocBuilder.scala:
```diff
@@ -110,6 +110,13 @@ trait DocBuilder {
     queryScope.addContent(new TablePlaceHolder(queryScope.assertions, queryScope.params:_*))
   }
 
+  def statsOnlyResultTable(): Unit = {
```
We should use this for all the empty parentheses in security (at least) to make at least this part is consistent

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityAdministrationTest.scala:
```diff
@@ -13,17 +13,46 @@ class SecurityAdministrationTest extends DocumentingTest with QueryStatisticsTes
     }
     section("Database administration", "administration-security-administration-database-privileges") {
-      p("")
+      synopsis("This section explains how to use Cypher to manage privileges for Neo4j database administrative rights.")
```
Should this section also contain all other privileges (like start and stop database)? right now it looks like its only contain access

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/admin-role-introduction.asciidoc:
```diff
@@ -0,0 +1,116 @@
+* _dbname_
+** The database to associate the privilege with.
+   Note that if you delete a database and create a new one with the same name, the new one will NOT have any of the privileges specifically assigned to the deleted graph.
+** Multiple databases can be specified, comma-separated.
```
This is not true - that doesn't parse

Hunterness replied:
Might be a bug in parsing though

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/grant-deny-syntax.asciidoc:
```diff
@@ -56,11 +54,9 @@ this combines both TRAVERSE and READ allowing an entity to be found and its prop
-   Note that if you delete a database and create a new one with the same name, the new one will NOT have any of the privileges specifically assigned to the deleted graph.
+** The graph or graphs to associate the privilege with. In {neo4j.version} there can be only one graph per database, and therefor this command uses the database name to refer to that graph.
+   Note that if you delete a database and create a new one with the same name, the new one will _NOT_ have any of the privileges specifically assigned to the deleted graph.
 ** Multiple graph names can be specified, comma-separated.
```
Not true, this does not parse

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/admin-role-introduction.asciidoc:
```diff
@@ -0,0 +1,116 @@
++
+allows indexes to be deleted on the specified database.
++
+** `INDEX MANAGEMENT`
```
The `INDEX MANAGEMENT`, `CONSTRAINT MANAGEMENT` and `NAME MANAGEMENT` has `MANAGEMENT` as optional
